navigation/insilico_rotater.h

```diff
@@ -16,76 +24,140 @@ namespace Navigation {
+                    rollImage(m_MaskImageOriginal, m_MaskImage, index);
+                }
             }
+        }
+
+        units::angle::radian_t columnToHeading(size_t column) const
+        {
+            return units::angle::turn_t{ (double) toIndex(column) / (double) m_ImageOriginal.cols };
+        }
+
+        size_t numRotations() const
+        {
+            return (m_EndRoll - m_BeginRoll) / m_ScanStep;
+        }
 
-            rollImage(m_Image, m_ScanStep);
-            if (!m_MaskImage.empty()) {
-                rollImage(m_MaskImage, m_ScanStep);
+    private:
+        const size_t m_ScanStep;
+        const IterType m_BeginRoll, m_EndRoll;
+        const cv::Mat &m_ImageOriginal, &m_MaskImageOriginal;
+        cv::Mat m_Image, m_MaskImage;
+
+        static void rollImage(const cv::Mat &imageIn, cv::Mat &imageOut, size_t pixels)
+        {
+            // Loop through rows
+            for (int y = 0; y < imageIn.rows; y++) {
+                // Get pointer to start of row
+                const uint8_t *rowPtr = imageIn.ptr(y);
+                uint8_t *rowPtrOut = imageOut.ptr(y);
+
+                // Rotate row to left by pixels
+                std::rotate_copy(rowPtr, rowPtr + pixels, rowPtr + imageIn.cols, rowPtrOut);
             }
         }
-    }
 
-    size_t max() const
+        static size_t distance(size_t first, size_t last)
+        {
+            return last - first;
+        }
+
+        template<typename Iter>
+        static size_t distance(Iter first, Iter last)
+        {
+            return static_cast<size_t>(std::distance(first, last));
+        }
+
+        static size_t toIndex(size_t index)
+        {
+            return index;
+        }
+
+        template<typename Iter>
+        static size_t toIndex(Iter it)
+        {
+            return *it;
+        }
+    };
+
+    template<typename IterType>
+    static auto
+    create(const cv::Size &unwrapRes,
+           const cv::Mat &maskImage,
+           const cv::Mat &image,
+           IterType beginRoll,
+           IterType endRoll,
+           size_t scanStep = 1)
     {
-        return (m_EndRoll - m_BeginRoll) / m_ScanStep;
+        return RotaterInternal<IterType>(unwrapRes, maskImage, image, beginRoll, endRoll, scanStep);
     }
 
-private:
-    const size_t m_ScanStep;
-    const size_t m_BeginRoll;
-    const size_t m_EndRoll;
-    cv::Mat m_Image, m_MaskImage;
+    static auto
+    create(const cv::Size &unwrapRes,
+           const cv::Mat &maskImage,
+           const cv::Mat &image,
+           size_t beginRoll,
+           size_t endRoll,
+           size_t scanStep = 1)
+    {
+        return RotaterInternal<size_t>(unwrapRes, maskImage, image, beginRoll, endRoll, scanStep);
+    }
 
-    static void rollImage(cv::Mat &image, size_t pixels)
+    static auto
+    create(const cv::Size &unwrapRes,
+           const cv::Mat &maskImage,
+           const cv::Mat &image,
+           size_t beginRoll = 0)
```

Comment:
It's because we want the default value of ``endRoll`` to be ``image.width`` so I couldn't use default arguments.

I did swap ``scanStep`` so it was at the end, but that was just because in the case where you're passing two iterators it seems odd to have to pass a ``scanStep`` of 1 beforehand.

Reply:
:+1: